---
title: Services
tags:
keywords: "features, capabilities, scalability, multichannel output, dita, hats, comparison, benefits"
last_updated: "July 20, 2017"
published: true
sidebar: site_sidebar
permalink: services.html
wide: true
toc: false
---

IIGB's bioinformatics facility offers several pipelined analysis services for NGS data. The Basic Read Processing is free of charge if performed as part of the NGS sequencing service. The downstream NGS-to-Spreadsheet analysis requires an extra service charge (see below for details). The latter is currently available for standard RNA-Seq, small RNA-Seq, SNP-Seq and ChIP-Seq samples that can be mapped against well annotated reference genomes. The following data processing steps and result reporting mechanisms are included in these services. Most of them use the systemPipeR NGS workflow and report generation environment available here.

Services | Description
--------|-----------
Basic Read Processing | Base calling, downloading of NGS files (FASTQ format), QC report, Demultiplexing upon user request
Data Analysis | RNA-Seq, Var-Seq, ChIP-Seq Methyl-Seq and others
Training | Instruction of hands-on tutorials and workshops on a wide variety of bioinformatics topics. Extensive manuals for these tutorials are available on our manual page.
Consulting | Establishment of research collaborations with experimental scientists from different departments.


## Data Analysis

Analysis | Description
--------|-----------|-----------
RNA-Seq| Differentially expressed genes. For details see RNA-Seq
Var-Seq| Variant calling and annotation. For details see Var-Seq
ChIP-Seq| Peak calling and annotation. For details see ChIP-Seq
Methyl-Seq| Methylated regions and annotation. For details see Methyl-Seq

Note: Customers are expected to provide a detailed outline of the design of their experiment. (e.g. treatments, replicates, comparisons), and the download information for the appropriate reference genome sequence and annotation data required for these analyses. For questions, please contact [Neerja Katiyar](mailto:neerja.katiyar@ucr.edu)

{% include links.html %}
