---
title: Past Workshops
tags:
  - getting_started
keywords: "features, capabilities, scalability, multichannel output, dita, hats, comparison, benefits"
last_updated: "July 20, 2017"
published: true
sidebar: mydoc_sidebar
permalink: mydoc_past_workshops.html
folder: mydoc
---

The bioinformatics core facility provides hands-on training workshops for the graduate students, postdocs, research scholars and faculty members.

Date | Time | Location | Instructor | Affiliation | Description | Material
August 4, 2017 | 10:00 - 12:00pm | University Laboratory Building Room 104 | Neerja Katiyar | UCR | Introduction to R Description - (1) Introduction to BigData (2) Rstudio environment (3) Data types, subsetting and filtering (4) Reading and writing a file in R (5) R graphics
June 23, 2017 | 9:00 - 12:00pm | University Laboratory Building Room 104 | Neerja Katiyar | UCR | Introduction to R Description - (1) R basics (2) Working with Rstudio environment (3) R graphics (4) Introduction to Bioconductor (5) Q & A (6) Discussion (Future workshop suggestions)
Dec 12, 2013 | 9:00am - 12:00pm	| Genomics Lecture Hall	| Thomas Girke	| UCR	| Introduction to R Description - R (http://www.r-project.org) is a versatile data analysis environment that has a broad application spectrum in all experimental and quantitative scientific areas. The associated Bioconductor project provides access to over 700 R extension packages for the analysis of modern biological and biomedical data sets, such as next generation sequences, comparative genomics, network modeling and statistical analysis. The R software is free and runs on all common operating systems. This workshop module provides an elementary-level introduction into the R environment to equip users with the knowledge required for the subsequent events of this R workshop series.The following topics will be covered (1) command syntax, (2) basic functions, (3) data import/export, (4) data, object types, (5) graphical display, (6) usage of R packages, libraries e.g. Bioconductor and (7) using R for basic data analysis operations. Schedule and Teaching Material Maximum number of participants - 75
Dec 12, 2013 | 1:00pm - 6:00pm	| Genomics Lecture Hall	| Thomas Girke	| UCR | Programming in R Description - In recent years the R language has become the Lingua Franca of data intensive research, and is now by far the most widely used data analysis programming language in bioinfomatics. One of the outstanding strengths of the R language is the ease of programming extensions to automate the analysis and mining of almost any data type. This workshop provides an overview of the basic knowledge for writing beginner level functions and programs in R. The following topics will be introduced: (1) conditional executions, (2) loops, (3) writing custom functions, (4) calling external software, (5) running and debugging R programs, and (6) building custom R packages. Basic knowledge of the R software, as introduced in the "Introduction to R" tutorial, will be expected in this workshop. Schedule and Teaching Material Maximum number of participants - 75
Dec 13, 2013 | 9:00am - 12:00pm	| Genomics Lecture Hall | Thomas Girke | UCR | Visualizing and Clustering High - Throughput Data with R Bioconducto Description - R is one of the most powerful environments for visualizing and clustering scientific data and creating beautiful publication quality graphics in a programmable and highly reproducible manner. This workshop will give an overview of the following topics (1) introduction to R base and grid graphics (2) usage of high-level graphics libraries including lattice and ggplot2 (3) data pre-processing for efficient visualization (4) writing of functions to generate customized graphics and automating image outputs (5) visualization of genome and next generation sequencing data with ggbio and Gviz (6) overview of the most common clustering algorithms used for profiling data, such as hierarchical clustering, fuzzy K-means clustering, principal component analysis, multidimensional scaling, biclustering and quality assessment of clustering results. Schedule and Teaching Material Maximum number of participants - 75
Dec 13, 2013 | 1:00pm - 4:00pm | Genomics Lecture Hall | Jordan Hayes & Thomas Girke | UCR | Linux Part I - Linux Essentials Description - The majority of bioinformatics software, especially in the next generation sequence analysis field, is only available for Unix/Linux-based operating systems. Basic knowledge about its usage provides free access to the most powerful and up-to-date applications for high-throughput data analysis. The workshop will teach beginners the basic command-line syntax for running applications on large data sets on LINUX systems. During the workshop users will work on IIGB's Linux cluster by logging in remotely from their laptops. The following topics will be covered: (1) overview of the Linux operating system, (2) file system organization, (3) getting around, (4) basic Shell commands and scripts, (5) available software, (6) running software like Bowtie, BWA, BLAST, HMMER, PHYLIP, EMBOSS, etc. Schedule and Teaching Material Maximum number of participants: 75
Dec 13, 2013 |	4:00pm - 6:00pm	| Genomics Lecture Hall	| Jordan Hayes & Thomas Girke	| UCR	| Linux Part II - Using IIGB's Linux Cluster Description - This seminar-style presentation will provide an introduction into the usage of the different load balancing and parallel computing tools available on IIGB's Linux cluster. A discussion will follow to determine the need for future hardware and software upgrades. Schedule and Teaching Material Maximum number of participants: 75
Dec 14, 2013 |	9:00am - 12:30pm | Genomics Lecture Hall | Thomas Girke	| UCR	| Basics on Analyzing Next Generation Sequencing Data with R/Bioconductor Description - R and Bioconductor provide extensive utilities for analyzing sequence data from traditional and next generation sequencing technologies (e.g. Sanger or Illumina). This workshop module will cover the  following topics: (1) basics on handling sequence, base call quality and annotation ranges in R (2) demultiplexing of pooled samples (3) generation of detailed quality reports of FASTQ files (4) quality/adaptor trimming and filtering of reads (5) parsing sequences by annotation ranges (6) handling alignment and coverage objects such as SAM/BAM files in R (7) interacting with external programs from R, such as short read aligners and peak callers (8) read count and density analysis and (9) genome visualization routines. Basic knowledge of the R software, as introduced in the "Introduction to R" tutorial, will be expected in this workshop. Schedule and Teaching Material Maximum number of participants: 75
Dec 14, 2013	| 1:30pm - 6:00pm |	Genomics Lecture Hall	| Thomas Girke	| UCR	| Analysis of RNA-Seq Data with R/Bioconductor Description - This workshop will cover the most common RNA-Seq data analysis routines. It will include the following topics (1) read mapping with intron/splice junction aware aligners (2) generation of read count data for genes, exons or other genome annotation ranges (3) normalization methods of read count data (4) statistical tests for identifying differentially expressed genes (DEGs) (5) enrichment analysis of GO terms and (7) visualization of read pileups using R graphics and the IGV genome browser and (6) enrichment analysis of GO terms. Basic knowledge of the R software including sequence handling routines, as introduced in the "Introduction to R" and Basic NGS tutorials, will be expected in this workshop. Schedule and Teaching Material Maximum number of participants: 75
Dec 15, 2013	| 9:00pm - 12:30pm	| Genomics Lecture Hall	| Thomas Girke	| UCR	| Analysis of ChIP-Seq Data with R/Bioconductor Description - This workshop will cover the most common ChIP-Seq data analysis routines. It will include the following topics: (1) short read alignment against reference genomes (2) efficient handling of read coverage data (3) peak calling with various algorithms implemented in R (4) integration of data from external peak callers (5) annotating peaks with genomic context information (5) statistical analysis of differential binding (6) peak viewing using R graphics, Gviz and the IGV genome browser (7) identification of enriched DNA motifs in peak sequences. Basic knowledge of the R software including sequence handling routines, as introduced in the "Introduction to R" and Basic NGS tutorials, will be expected in this workshop. Schedule and Teaching Material Maximum number of participants: 75 
Dec 15, 2013	| 1:30pm - 6:00pm	| Genomics Lecture Hall	| Thomas Girke	| UCR	| Analysis of SNP/Var-Seq Data with R/Bioconductor Description - This workshop will cover the most common SNP/Var-Seq data analysis routines. It will include the following topics: (1) read mapping with variant aware aligners (e.g. BWA, GSNAP) (2) SNP/indel calling (VariantTools, GATK, BCFtools) (3) handling of standard variant data formats such as VCF (4) annotating variants with genomic context information including variant mapping to genes, intergenic regions (5) identification of synonymous/non-synonymous SNPs (6) variant viewing using R graphics (ggbio) and the IGV genome browser. Basic knowledge of the R software including sequence handling routines, as introduced in the "Introduction to R" and Basic NGS tutorials, will be expected in this workshop. Schedule and Teaching Material Maximum number of participants: 75
Dec 16, 2013	| 9:00am - 12:00am	| Genomics Lecture Hall	| Tyler Backman & Thomas Girke	| UCR	| Analysis of Drug-like Small Molecules and High-Throughput Screens with R/Bioconductor Description - This workshop introduces various R packages useful for analyzing drug-like compound and screening data sets. This includes the packages ChemmineR, ChemmineOB, fmcsR, eiR and bioassayR. Efficient R functions will be introduced for handling/analyzing SDF/MOL files, interfacing with PubChem, structural similarity searching, clustering of compound libraries with a wide spectrum of algorithms and utilities for managing complex compound data sets. In addition, visualization functions for compound clusters and chemical structures will be introduced. The last part will cover the analysis of large publicly available high-throughput screening data sets like PubChem BioAssay. Basic knowledge of the R software, as introduced in the "Introduction to R" tutorial, will be expected in this workshop. Schedule and Teaching Material Maximum number of participants: 75
Dec 16, 2013	| 1:30pm - 5:30pm	| Genomics Lecture Hall	| Neerja Katiyar & Thomas Girke	| UCR	| Web-based Analysis of Next Generation Sequence Data Description - This workshop will introduce basic NGS data analysis routines using the web-based Galaxy environment. It will include the following topics: (1) assessment of read qualities (2) read trimming and filtering routines (3) aligning reads to reference genomes (4) generation of read counts for RNA-Seq data (5) peak calling for ChIP-Seq experiments and (6) visualization routines of read pileups along with annotation information using the freely and very easy-to-use IGV genome browser from the Broad Institute. The material will be useful for both complete beginners and intermediate users (e.g. attended previous R workshop on NGS data analysis). No special computer knowledge is required for this workshop. Teaching Material Maximum number of participants: 75
July 18, 2013	| 2:00pm - 5:00am	| Fred Hutchinson Cancer Research Center - Seattle, WA	| Thomas Girke	| UCR	| Cheminformatics of Drug-like Small Molecules Description - This lab session will introduce several Bioconductor packages (ChemmineR, fmcsR and eiR) for analyzing drug-like small molecule and high-throughput screening data in R. These packages contain utilities for efficient processing of large numbers of molecules, physicochemical/structural property predictions, structural similarity searching, classification and clustering of compound screening libraries, and bioactivity data with a wide spectrum of algorithms. In addition, they offer visualization functions for compound clusters and chemical structures. Teaching Material Maximum number of participants: 75

Note: Customers are expected to provide a detailed outline of the design of their experiment. (e.g. treatments, replicates, comparisons), and the download information for the appropriate reference genome sequence and annotation data required for these analyses. For questions, please Contact Us.

{% include links.html %}
